Completion of Site Creation wizard

stevebaranski opened the pull request:
### Description

Begins to address #10462, which calls for the transition from a wizard flow to invocation of the site creation service, with the output that's accumulated.

### Testing

- The branch should build & tests (both old & new) should pass.
- Navigating through the _Enhanced Site Creation_ should through the flow to the final _Success_ screen. Results should match the exhibits below.

### Screenshots

_iPhone SE_

![10462_se_02](https://user-images.githubusercontent.com/221062/49915257-71327500-fe49-11e8-9000-d0e12aea30e9.gif)

_iPad Portrait_
<img width="878" alt="10462_ipad_portrait" src="https://user-images.githubusercontent.com/221062/49915270-7e4f6400-fe49-11e8-9414-2c6ea7d04d8f.png">

_iPad Landscape_
<img width="1109" alt="10462_ipad_landscape" src="https://user-images.githubusercontent.com/221062/49915272-827b8180-fe49-11e8-8301-dd614f95e8e6.png">

stevebaranski commented:
@ctarda as with #10604, I created this branch off of the one for #10603, so there are some redundant commits noted here. For this reason, I affixed the `DO NOT MERGE` tag unless or until that PR has been merged.

You should be able to focus your review on the last three commits.

I wanted to use this PR as the basis for discussion for how we complete the wizard flow. I think we need to iron out some of the relationships a bit. Namely:

- The flow is initiated with the `SiteCreationWizardLauncher`, which instantiates the `SiteCreator`, the `WizardSteps`, and the `SiteCreationWizard` itself.
- In this PR, I added a `complete` step to the `WizardDelegate` - is this what you had in mind?
- We need to settle on how we conclude the wizard and handle that sequence appropriately. We've accumulated the data in `SiteCreator`, and now we need to determine _where_ the `SiteCreatorOutput` is generated. Please feel free to comment on the PR and I'll amend it iteratively, or you can push commits to the branch that reflect your thinking.

Once we have consensus, we can merge these changes and I can proceed with #10462 & #10429.

ctarda commented:
> In this PR, I added a complete step to the WizardDelegate 
That would imply that the final step of the wizard would need to be aware of the fact that it is the final step, which could be argued that implies that all steps need to be aware of the fact that they are or are not the final one. 

Would it make sense, to avoid having the steps to know their position in the wizard, to let `WizardNavigation` decide if the wizard has been completed?

Also, if there is any particular UI that needs to be rendered while the `SiteCreatorOutput` is committed to the backend, would it make sense to encapsulate that UI as another step? That way, it would not even be necessary for `WizardNavigation` to check if the wizard has been completed

stevebaranski commented:
@SylvesterWilmott I have attached an animated GIF and screenshots of how the final step in the site creation flow look on a range of devices. Zeplin didn't provide a lot of detail on iPad - is this okay? If not, I'm happy to revise this based on your input.

stevebaranski commented:
@ctarda thanks for your earlier review & input. I have taken that to heart and pushed additional commits that substantially complete this issue. For clarity, it includes the following changes:

- I removed the method I proposed earlier to the `Wizard` protocol.
- I deleted `SiteCreationWizardTitle` as it wasn't being used.
- I organized our site creation tests, and added some additional ones.
- I implemented the _Loading / Success_ transition as depicted in Zeplin; it uses a mock service & is currently hard-wired to render the _Longreads_ website.
- I added the `[Status] Needs Design Review` to welcome Sylvester's input.

Would you be able to take a look and let me know what you think?

ctarda commented on WordPress/Classes/ViewRelated/Site Creation/FinalAssembly/SiteAssemblyContentView.swift:
```diff
@@ -0,0 +1,282 @@
+    // MARK: SiteAssemblyContentView
+
+    init() {
+        self.completionLabel = {
```
I don't see this is as a blocker, but I find this way of creating the properties makes the initialiser difficult to parse, as it turns into a long method with several long blocks of indented code. Again, my personal preference, and not a blocker, but would you consider declaring these properties as lazy var? (I still like a lot the idea of declaring in a closure)

ctarda commented on WordPress/Classes/ViewRelated/Site Creation/FinalAssembly/SiteAssemblyContentView.swift:
```diff
@@ -0,0 +1,282 @@
+/// This view is intended for use as the root view of `SiteAssemblyWizardContent`.
+/// It manages the state transitions that occur as a site is assembled via remote service dialogue.
+///
+class SiteAssemblyContentView: UIView {
```
Do we really want this class to not be final?

stevebaranski replied:
Thanks for calling that out. I have pushed an additional commit that marks both `AssembledSiteView` & `SiteAssemblyContentView` as `final`.

stevebaranski commented:
@ctarda thanks for the valuable input & approval. I have addressed your comments, and will remove the `[Status] DO NOT MERGE` from the PR.

I will proceed to merge these changes once I have approval from Sylvester.

stevebaranski commented:
@SylvesterWilmott thanks for the this input! I have pushed a few additional commits that should address the comments you had. I hope you'll find this more to your liking.

_iPhone SE_
<img width="504" alt="10462-revised-iphone-se" src="https://user-images.githubusercontent.com/221062/49973434-e5bcf080-fee9-11e8-9483-51c4f6adc683.png">

_iPad Pro (Portrait)_
<img width="842" alt="10462-revised-ipad-portrait" src="https://user-images.githubusercontent.com/221062/49973438-e81f4a80-fee9-11e8-993b-efb8a3883822.png">

_iPad Pro (Landscape)_
<img width="1110" alt="10462-revised-ipad-landscape" src="https://user-images.githubusercontent.com/221062/49973443-ec4b6800-fee9-11e8-8c84-533f1d98b5b7.png">

SylvesterWilmott commented:
Looks really good! Thanks @stevebaranski
